### PR TITLE
1664: Untouched State

### PR DESCRIPTION
**Pull request by f1sh1918**
### Short description

As a user i don't want to get initial error message if i did not interact with the form

### Proposed changes

- add untouched states for self service form
- add sendAttempt state and pass them to extensions to also show errors if a send attempts was made without filling required fields

### Side effects

- none

### Testing

1. Open /erstellen
2. No error messages should be shown initially to the user
3. Click "Koblenz Pass erstellen" without filling anything or not filling all required fields. All fields should be validated and show error messages
4. Test if all fields still will be validated correctly

### Resolved issues

Fixes: #1664

### CODE REVIEW

File: administration/src/bp-modules/self-service/CardSelfServiceForm.tsx [hunk 1]
```diff
@@ -43,12 +43,27 @@ const CardSelfServiceForm = ({
...
   const [openDataPrivacy, setOpenDataPrivacy] = useState<boolean>(false)
   const [openReferenceInformation, setOpenReferenceInformation] = useState<boolean>(false)
   const cardValid = isValid(card, { expirationDateNullable: true })
   const appToaster = useAppToaster()
 
+  const updateFullName = (fullName: string) => {
+    updateCard({ fullName })
+    if (!touchedFullName) {
+      setTouchedFullName(true)
+    }
```
Comment (steffenkleinle) [inline administration/src/bp-modules/self-service/CardSelfServiceForm.tsx:57]: :wrench: If this is named like this, I think we should always set `touchedFullName` to `true` on updates.

File: administration/src/bp-modules/self-service/CardSelfServiceForm.tsx [hunk 2]
```diff
@@ -43,12 +43,27 @@ const CardSelfServiceForm = ({
...
+  }
+
   const createKoblenzPass = async () => {
+    if (!formSendAttempt) {
+      setFormSendAttempt(true)
+    }
+    if (!touchedFullName) {
+      setTouchedFullName(true)
+    }
```
Comment (steffenkleinle) [inline administration/src/bp-modules/self-service/CardSelfServiceForm.tsx:66]: :wrench: The checks are not needed, if the state is already true before, nothing happens (as the value does not change).
Also, it should be enough to only set `formSendAttempt` to true.
```suggestion
    setFormSendAttempt(true)
```

File: administration/src/bp-modules/self-service/CardSelfServiceForm.tsx [hunk 3]
```diff
@@ -80,13 +95,13 @@ const CardSelfServiceForm = ({
...
               />
             }
-            intent={isFullNameValid(card) ? undefined : Intent.DANGER}
+            intent={isFullNameValid(card) || !touchedFullName ? undefined : Intent.DANGER}
             value={card.fullName}
-            onChange={event => updateCard({ fullName: event.target.value })}
+            onChange={event => updateFullName(event.target.value)}
           />
-          <FormErrorMessage errorMessage={getFullNameValidationErrorMessage(card.fullName)} />
+          {touchedFullName && <FormErrorMessage errorMessage={getFullNameValidationErrorMessage(card.fullName)} />}
```
Comment (steffenkleinle) [inline administration/src/bp-modules/self-service/CardSelfServiceForm.tsx:102]:
```suggestion
          {showErrorMessage && <FormErrorMessage errorMessage={getFullNameValidationErrorMessage(card.fullName)} />}
```

File: administration/src/cards/extensions/BirthdayExtension.tsx [hunk 4]
```diff
@@ -24,20 +31,24 @@ const BirthdayForm = ({ value, setValue, isValid }: ExtensionComponentProps<Birt
...
+  }
 
   return (
     <FormGroup label='Geburtsdatum'>
       <CustomDatePicker
         date={birthday?.toLocalDate() ?? null}
         onChange={changeBirthday}
         onClear={() => setValue({ birthday: null })}
-        isValid={isValid}
+        isValid={isValid || (!touched && !showRequired)}
```
Comment (steffenkleinle) [inline administration/src/cards/extensions/BirthdayExtension.tsx:47]:
```suggestion
        isValid={isValid || !showErrorMessage}
```

File: administration/src/cards/extensions/BirthdayExtension.tsx [hunk 5]
```diff
@@ -24,20 +31,24 @@ const BirthdayForm = ({ value, setValue, isValid }: ExtensionComponentProps<Birt
     return null
   }
 
-  const changeBirthday = (date: Date | null) =>
+  const changeBirthday = (date: Date | null) => {
     setValue({ birthday: PlainDate.safeFromCustomFormat(date?.toLocaleDateString() ?? null) })
+    if (!touched) {
+      setTouched(true)
+    }
```
Comment (steffenkleinle) [inline administration/src/cards/extensions/BirthdayExtension.tsx:38]:
```suggestion
    setTouched(true)
```
Same for other occurrences.

File: administration/src/bp-modules/self-service/CardSelfServiceForm.tsx [hunk 6]
```diff
@@ -80,13 +95,13 @@ const CardSelfServiceForm = ({
                 input={card.fullName}
               />
             }
-            intent={isFullNameValid(card) ? undefined : Intent.DANGER}
+            intent={isFullNameValid(card) || !touchedFullName ? undefined : Intent.DANGER}
```
Comment (steffenkleinle) [inline administration/src/bp-modules/self-service/CardSelfServiceForm.tsx:98]:
```suggestion
const showErrorMessage = touchedFullName || formSendAttempt
            intent={isFullNameValid(card) || !showErrorMessage ? undefined : Intent.DANGER}
```

File: administration/src/cards/extensions/KoblenzReferenceNumberExtension.tsx [hunk 7]
```diff
@@ -20,38 +20,49 @@ const KoblenzReferenceNumberExtensionForm = ({
...
-      intent={isValid ? undefined : Intent.DANGER}>
+    <FormGroup label='Referenznummer' labelFor='koblenz-reference-number-input'>
       <InputGroup
         fill
         large={viewportSmall}
         id='koblenz-reference-number-input'
         placeholder='5.012.067.281, 000D000001, 99478'
-        intent={isValid ? undefined : Intent.DANGER}
-        value={value.koblenzReferenceNumber}
+        intent={isValid || (!touched && !showRequired) ? undefined : Intent.DANGER}
```
Comment (steffenkleinle) [inline administration/src/cards/extensions/KoblenzReferenceNumberExtension.tsx:58]:
```suggestion
        intent={isValid || !showErrorMessage ? undefined : Intent.DANGER}
```

**Review by steffenkleinle — COMMENTED** [review]
Personally, I think errors should only be shown once the user leaves the field (or the send button is clicked). Otherwise it always feels a little aggressive if I type the first letter of my name and already get an error message. Not sure if that is easily possible though.

Do you plan to implement this for the other extensions/expiration date for the other passes/cards as well?

Otherwise works as expected, nice work.

**Comment by f1sh1918** [comment]
> Personally, I think errors should only be shown once the user leaves the field (or the send button is clicked). Otherwise it always feels a little aggressive if I type the first letter of my name and already get an error message. Not sure if that is easily possible though. This would also mitigate the issue that `onChange` is not called if just leaving the field again without any input.
> 
> Do you plan to implement this for the other extensions/expiration date for the other passes/cards as well?
> 
> Otherwise works as expected, nice work.

I see your point that it looks aggressive. On the other hand if you are on a mobile device and type in a name that has issues and you have to click to the next field to see the issue to then return to the name field to solve it. This is also quite annoying

**Review by steffenkleinle — APPROVED** [review]
not tested again

**Pull request merged**